cpp/src/arrow/ipc/reader.h

```diff
@@ -190,6 +193,85 @@ class ARROW_EXPORT RecordBatchFileReader {
   virtual Status ReadRecordBatch(int i, std::shared_ptr<RecordBatch>* batch) = 0;
 };
 
+/// \class RecordBatchReceiverCollect
+/// \brief Collect record batches read by RecordBatchStreamEmitter.
+class ARROW_EXPORT RecordBatchReceiverCollect : public Receiver {
+ public:
+  RecordBatchReceiverCollect() : record_batches_() {}
+  virtual ~RecordBatchReceiverCollect() = default;
+
+  Status RecordBatchReceived(std::shared_ptr<RecordBatch> record_batch) override {
+    record_batches_.push_back(std::move(record_batch));
+    return Status::OK();
+  }
+
+  /// \return the all read record batches
+  std::vector<std::shared_ptr<RecordBatch>> record_batches() const {
+    return record_batches_;
+  }
+
+ private:
+  std::vector<std::shared_ptr<RecordBatch>> record_batches_;
+};
+
+/// \class RecordBatchStreamEmitter
+/// \brief Push style record batch stream reader that receives data
+/// from user.
+///
+/// This class reads the schema (plus any dictionaries) as the first
+/// messages in the fed data, followed by record batches.
+class ARROW_EXPORT RecordBatchStreamEmitter {
+ public:
+  /// \brief Construct a record batch emitter.
+  ///
+  /// \param[in] receiver a Receiver that must implement
+  /// Receiver::MessageReceived to receive read record batches
+  /// \param[in] options any IPC reading options (optional)
+  RecordBatchStreamEmitter(Receiver* receiver,
+                           const IpcReadOptions& options = IpcReadOptions::Defaults());
+
+  virtual ~RecordBatchStreamEmitter();
+
+  /// \brief Feed data to the emitter as a raw data.
+  ///
+  /// If the emitter can read one or more record batches by the data,
+  /// the emitter emits read record batches by calling
+  /// receiver->Receive() multiple times.
+  ///
+  /// \param[in] data a raw data to be processed. This data isn't
+  /// copied. The passed memory must be kept alive through record
+  /// batch processing.
+  /// \param[in] size raw data size.
+  /// \return Status
+  Status Consume(const uint8_t* data, int64_t size);
+
+  /// \brief Feed data to the emitter as a Buffer.
+  ///
+  /// If the emitter can read one or more record batches by the
+  /// Buffer, the emitter emits read record batches by calling
+  /// receiver->Receive() multiple times.
+  ///
+  /// \param[in] buffer a Buffer to be processed.
+  /// \return Status
+  Status Consume(std::shared_ptr<Buffer> buffer);
+
+  /// \return the shared schema of the record batches in the stream
+  std::shared_ptr<Schema> schema() const;
```

Comment:
Instead, perhaps a `SchemaReceived` method on `Receiver`? With a default implementation returning `Status::OK`, since that is not always necessary to process.